### PR TITLE
Update zoneprinter code to include more functionality

### PR DESCRIPTION
**Pull request by rramkumar1**
New functionality includes:

1. Configurable health check path + port
2. ability to kill health checks (i.e simulate health check failures)
3. Configurable version to simulate deployment upgrades.
4. root path which dumps useful information including version

Note that these changes introduce no new dependencies.

---
This change is [Reviewable](https://reviewable.io/reviews/googlecloudplatform/k8s-multicluster-ingress/235)

### CODE REVIEW

File: examples/zone-printer/src/main.go [hunk 1]
```diff
@@ -38,13 +56,74 @@ func main() {
...
 
-	log.Println("starting to listen on port 80")
-	http.HandleFunc("/", handle)
-	err := http.ListenAndServe(":80", nil)
-	log.Fatal(err)
+	finished := make(chan bool)
+
+	log.Printf("starting to listen on port %d", mainPort)
+	mainServer := http.NewServeMux()
+	mainServer.HandleFunc("/", handlePing)
```
Comment (ahmetb) [inline examples/zone-printer/src/main.go:63]: please point root to `handleLocation` handler. this app is supposed to print zone mainly. anything else is secondary.

File: examples/zone-printer/src/main.go [hunk 2]
```diff
@@ -38,13 +56,74 @@ func main() {
...
+	Hostname string
+	Version string
+	GCPZone string
+	Backend string
+}
+
+func handlePing(w http.ResponseWriter, r *http.Request) {
+	podHostname, err := os.Hostname()
+	if err != nil {
+		http.Error(w, err.Error(), http.StatusInternalServerError)
```
Comment (ahmetb) [inline examples/zone-printer/src/main.go:93]: you're missing a return here?

File: examples/zone-printer/src/main.go [hunk 3]
```diff
@@ -38,13 +56,74 @@ func main() {
 		log.Printf("info: determined zone: %q", zone)
 	}
 
-	log.Println("starting to listen on port 80")
-	http.HandleFunc("/", handle)
-	err := http.ListenAndServe(":80", nil)
-	log.Fatal(err)
+	finished := make(chan bool)
```
Comment (ahmetb) [inline examples/zone-printer/src/main.go:59]: this chan seems unused? promote the main handler to outside the goroutine perhaps?

File: examples/zone-printer/src/main.go [hunk 4]
```diff
@@ -38,13 +56,74 @@ func main() {
...
+	lock.Lock()
+	defer lock.Unlock()
+
+	killed = !killed
+	fmt.Fprintf(w, "Successfully toggled kill status")
+}
+
+func handleHealthCheck(w http.ResponseWriter, r *http.Request) {
+	lock.Lock()
+	defer lock.Unlock()
```
Comment (ahmetb) [inline examples/zone-printer/src/main.go:117]: bool’s are atomic so you actually don't need locks around them.

if you're really interested in locking you can prob. use RWLock and gain a RLock here.

File: examples/zone-printer/src/main.go (hunk 2 above)
Comment (rramkumar1) [reply examples/zone-printer/src/main.go:87]: Done

File: examples/zone-printer/src/main.go (hunk 1 above)
Comment (rramkumar1) [reply examples/zone-printer/src/main.go]: Done

File: examples/zone-printer/src/main.go (hunk 3 above)
Comment (rramkumar1) [reply examples/zone-printer/src/main.go]: Done

File: examples/zone-printer/src/main.go (hunk 2 above)
Comment (rramkumar1) [reply examples/zone-printer/src/main.go:87]: Done

File: examples/zone-printer/src/main.go (hunk 4 above)
Comment (rramkumar1) [reply examples/zone-printer/src/main.go]: Really? I don't see any documentation on that. Used RWMutex.

**Comment by ahmetb** [comment]
Please make sure you also run gofmt, I also notice a compilation error in handlefunc for /.

**Comment by rramkumar1** [comment]
@ahmetb Should be fixed now.

**Comment by nikhiljindal** [comment]
lgtm

Will wait for @ahmetb's  lgtm before merging.

**Pull request merged**